src/compute/src/render/reduce.rs

```diff
@@ -317,92 +319,138 @@ where
+                    // equi-join on the key, similarly to the proposal in PR #17013.
+                    //
+                    // Note that we also do not want to have anything left over to stich.  If we do,
+                    // then we also have an error and would violate join semantics.
+                    if accumulable.next() == None
+                        && hierarchical.next() == None
+                        && basic.next() == None
+                    {
+                        output.push((row_buf.clone(), 1));
                     }
-                } else {
-                    // We expected to stitch together exactly as many aggregate types
-                    // as requested by the collation. If we cannot, we log an error and
-                    // produce no output for this key.
-                    warn!("[customer-data] Aggregates in input differ ({}) from requested ({}) for key in ReduceCollation: {key} in dataflow {debug_name}",
-                        input.len(), distinct_aggregate_types.len());
-                    soft_assert_or_log!(false, "Mismatched aggregates for key in ReduceCollation");
                 }
-            }
-        })
+            },
+            move |key, input, output| {
+                if input.len() != n_distinct_aggregate_types {
+                    // We expected to stitch together exactly as many aggregate types as requested
+                    // by the collation. If we cannot, we log an error and produce no output for
+                    // this key.
+                    let message = "Mismatched aggregates for key in ReduceCollation";
+                    warn!(
+                        ?key,
+                        debug_name,
+                        n_aggregates_requested = input.len(),
+                        n_distinct_aggregate_types,
+                        "[customer-data] {message}"
+                    );
+
+                    error!("{message}");
+                    output.push((EvalError::Internal(message.to_string()).into(), 1));
+                    return;
+                }
+
+                let mut accumulable = DatumList::empty().iter();
+                let mut hierarchical = DatumList::empty().iter();
+                let mut basic = DatumList::empty().iter();
+                for ((reduction_type, row), _) in input.iter() {
+                    match reduction_type {
+                        ReductionType::Accumulable => accumulable = row.iter(),
+                        ReductionType::Hierarchical => hierarchical = row.iter(),
+                        ReductionType::Basic => basic = row.iter(),
+                    }
+                }
+
+                for typ in aggregate_types_err.iter() {
+                    let datum = match typ {
+                        ReductionType::Accumulable => accumulable.next(),
+                        ReductionType::Hierarchical => hierarchical.next(),
+                        ReductionType::Basic => basic.next(),
```

Comment:
Calling `next` on a `DatumList` can be expensive because it needs to decode the datum. Unfortunately, I don't see a way to avoid this as we don't have any way to just indicate whether a datum exists, without decoding it. We should at least try to inline `read_datum` to enable dead-code analysis...

Reply:
At the same time, `read_datum` might be one function we _don't_ want to inline, because it's a massive beast. Maybe we should have a special iterator that yields tags/bools instead of datums?